### PR TITLE
Improve displacement mapping

### PR DESCRIPTION
**Pull request by RealBadAngel**

### CODE REVIEW

File: client/shaders/nodes_shader/opengl_fragment.glsl [hunk 1]
```diff
@@ -19,80 +20,135 @@ bool normalTexturePresent = false;
...
+	}
+	depth = best_depth - size;
+	for (int i = 0 ; i < binary_steps ; ++i) {
+		size *= 0.5;
+		vec4 t = texture2D(normalTexture, dp + ds * depth);
+		if (depth >= t.a) {
+			best_depth = depth;
+			depth -= 2 * size;
+		}
+	depth += size;
```
Comment (rubenwardy) [inline client/shaders/nodes_shader/opengl_fragment.glsl]: indentation here :P

Comment (RealBadAngel) [inline client/shaders/nodes_shader/opengl_fragment.glsl]: where???

File: src/mapblock_mesh.cpp [hunk 2]
```diff
@@ -1480,25 +1475,22 @@ void MeshCollector::append(const TileSpec &tile,
...
 		pp.tile = tile;
 		prebuffers.push_back(pp);
-		p = &prebuffers[prebuffers.size()-1];
+		p = &prebuffers[prebuffers.size() - 1];
 	}
 
 	u32 vertex_count = p->vertices.size();
-	for(u32 i=0; i<numIndices; i++)
-	{
+	for (u32 i = 0; i < numIndices; i++)	{
```
Comment (est31) [inline src/mapblock_mesh.cpp]: space no tab

File: client/shaders/nodes_shader/opengl_fragment.glsl [hunk 3]
```diff
@@ -19,80 +20,135 @@ bool normalTexturePresent = false;
...
+		if (depth >= t.a) {
+			best_depth = depth;
+			depth -= 2 * size;
+		}
+		depth += size;
+	}
+	return best_depth;
+}
+
+float find_intersectionRGB(vec2 dp, vec2 ds) {
```
Comment (est31) [inline client/shaders/nodes_shader/opengl_fragment.glsl]: newline

File: client/shaders/nodes_shader/opengl_vertex.glsl [hunk 4]
```diff
@@ -14,6 +14,7 @@ varying vec3 eyeVec;
 varying vec3 lightVec;
 varying vec3 tsEyeVec;
 varying vec3 tsLightVec;
+varying float generate_heightmaps;
```
Comment (est31) [inline client/shaders/nodes_shader/opengl_vertex.glsl]: Why is this a float and no int?

File: client/shaders/water_surface_shader/opengl_fragment.glsl [hunk 5]
```diff
@@ -62,8 +62,7 @@ void main (void)
 	} 
 #endif
 		
-#ifdef GENERATE_NORMALMAPS
-	if (use_normalmap == false){
+	if (GENERATE_NORMALMAPS == 1 && use_normalmap == false){
```
Comment (est31) [inline client/shaders/water_surface_shader/opengl_fragment.glsl]: space :)

File: client/shaders/nodes_shader/opengl_fragment.glsl [hunk 6]
```diff
@@ -19,80 +20,135 @@ bool normalTexturePresent = false;
...
-		vec3 tsEye = normalize(tsEyeVec);
-		float height = PARALLAX_OCCLUSION_SCALE * texture2D(normalTexture, uv).a - PARALLAX_OCCLUSION_BIAS;
-		uv = uv + texture2D(normalTexture, uv).z * height * vec2(tsEye.x,-tsEye.y);
+	vec3 eyeRay = normalize(tsEyeVec);	
+#if PARALLAX_OCCLUSION_MODE == 0
+	// Parallax occlusion with slope information
+	if (normalTexturePresent) {
+		const float scale = PARALLAX_OCCLUSION_SCALE / PARALLAX_OCCLUSION_ITERATIONS;
+		const float bias = PARALLAX_OCCLUSION_BIAS / PARALLAX_OCCLUSION_ITERATIONS;
+		for(int i = 0; i < PARALLAX_OCCLUSION_ITERATIONS; i++) {
```
Comment (est31) [inline client/shaders/nodes_shader/opengl_fragment.glsl]: space :)

File: src/mapblock_mesh.cpp [hunk 7]
```diff
@@ -1158,20 +1159,21 @@ MapBlockMesh::MapBlockMesh(MeshMakeData *data, v3s16 camera_offset):
...
 		{
+			video::S3DVertexTangents *vertex = &p.vertices[j];
 			// Note applyFacesShading second parameter is precalculated sqrt
 			// value for speed improvement
 			// Skip it for lightsources and top faces.
-			video::SColor &vc = p.vertices[j].Color;
+			video::SColor &vc = vertex->Color;
 			if (!vc.getBlue()) {
-				if (p.vertices[j].Normal.Y < -0.5) {
+				if (vertex->Normal.Y < -0.5) {
```
Comment (est31) [inline src/mapblock_mesh.cpp]: perhaps cache `vertex->Normal` too?

Comment (RealBadAngel) [inline src/mapblock_mesh.cpp]: that would make code less sensible

Comment (est31) [inline src/mapblock_mesh.cpp]: you mean readable?

**Comment by est31** [comment]
43fcfbfe05578d7471d40c8c087fd04e24b264b5

**Pull request closed without merging**